### PR TITLE
Update events.md

### PR DESCRIPTION
**Pull request by ChrisThrasher**
See #193

### CODE REVIEW

File: pages/tutorials/3.0/window/events.md [hunk 1]
```diff
@@ -7,259 +7,329 @@
...
 ## Introduction
 
-This tutorial is a detailed list of window events. It describes them, and shows how to (and how not to) use them.
+This tutorial is a detailed list of window events.
+It describes them, and shows how to (and how not to) use them.
 
 ## The sf::Event type
 
-Before dealing with events, it is important to understand what the [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") type is, and how to correctly use it. [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") is a _union_, which means that only one of its members is valid at a time (remember your C++ lesson: all the members of a union share the same memory space). The valid member is the one that matches the event type, for example `event.key` for a `KeyPressed` event. Trying to read any other member will result in an undefined behavior (most likely: random or invalid values). It is important to never try to use an event member that doesn't match its type.
+SFML uses `std::variant` under the hood to implement a type-safe API for events.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:15]: This should IMHO not be the opening statement of the event tutorial. People land here because they want to understand how to use the API, if they wanted to know the implementation, they'd go to GitHub 😄

File: pages/tutorials/3.0/window/events.md [hunk 2]
```diff
@@ -7,259 +7,329 @@
...
+There are two main ways to use this API.
+Check out the EventHandling example program to see these methods in practice.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:17]:
```suggestion
For a comprehensive example of both approaches, check out the [EventHandling example program](https://github.com/SFML/SFML/tree/master/examples/event_handling "EventHandling example on GitHub"].
```

File: pages/tutorials/3.0/window/events.md [hunk 3]
```diff
@@ -7,259 +7,329 @@
...
 
-To be clear, here is what a typical event loop looks like:
+The first option is based around `sf::Event::getIf<T>` and `sf::Event::is<T>`.
+`getIf<T>` works by providing a template parameter which must be an event subtype.
+Event subtypes are types like `sf::Event::Resized` or `sf::Event::MouseMoved`.
+If the template argument matches the active event subtype then a pointer to the data is returned.
+If that template argument is not the active event subtype then `nullptr` is returned.
+`sf::Event::is<T>` more simply returns `true` if `T` matches the active event subtype.
+`is<T>` is often used for subtypes like `sf::Event::Closed` which contain no data.
+Here's what that looks like:
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:28]: I find the text rather hard to follow. Even though I know the new API, I had to re-read the fourth line three times before I understood everything.

What do you think of the following approach?

```suggestion
The first option is based around `sf::Event`'s member functions `getIf<T>` and `is<T>`.
The template type `T` is an event subtype, such as `sf::Event::Resized` or `sf::Event::MouseMoved`.

`getIf<T>` returns a pointer to the subtype specific event data, if the provided template type matches the active event subtype, otherwise `nullptr` is returned.

`is<T>` returns `true` if the provided template type matches the active event subtype, `false` otherwise.
This is especially useful for subtypes like `sf::Event::Closed` which contain no data.
Here's what that looks like:
```

File: pages/tutorials/3.0/window/events.md [hunk 4]
```diff
@@ -7,259 +7,329 @@
...
+        }
     }
+
+    // Remainder of main loop
 }
 ```
 
-Read the above paragraph once again and make sure that you fully understand it, the [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") union is the cause of many problems for inexperienced programmers.
+Note how the API for getting events has changed slightly.
+`sf::WindowBase::pollEvent` and `sf::WindowBase::waitEvent` now return a `std::optional<sf::Event>`.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:51]: I think this should be at the topic, similar to how it was before. It makes sense to me, that we first show how to get events and then explain what you can do with them.

File: pages/tutorials/3.0/window/events.md [hunk 5]
```diff
@@ -7,259 +7,329 @@
...
+These two functions _might_ return an event but they might not.
+C++ lets you deduce the template parameter which is why you can write `const std::optional event` instead of `const std::optional<sf::Event> event`.
+`const auto event` is another valid choice if you prefer a more terse expression.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:54]: I'd opt for "simple English" terms.
```suggestion
`const auto event` is another valid choice if you prefer a shorter expression.
```

File: pages/tutorials/3.0/window/events.md [hunk 6]
```diff
@@ -7,259 +7,329 @@
...
+
+### `sf::WindowBase::handleEvents`
 
-Alright, now we can see what events SFML supports, what they mean and how to use them properly.
+The second option for processing events is via the `sf::WindowBase::handleEvents` function.
+This functions performs event visitation.
+What this means is that you can provide callbacks which take different event subtypes as arguments.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:60]:
```suggestion
What this means is that you can provide lambdas or other callbacks which take different event subtypes as arguments.
```

File: pages/tutorials/3.0/window/events.md [hunk 7]
```diff
@@ -7,259 +7,329 @@
...
+Alternatively you may provide an object (or objects) with `operator()` implementations which handle the event subtypes you want to process.
+Notably you do not have to provide callbacks for all possible event subtypes.
+Depending on the current active event subtype, the corresponding callback is called.
+Here's what that looks like:
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:64]:
```suggestion
Depending on the current active event subtype, the corresponding callback is called.

!!! note

    You do not have to provide callbacks for all possible event subtypes.

Here's what that looks like:
```

File: pages/tutorials/3.0/window/events.md [hunk 8]
```diff
@@ -7,259 +7,329 @@
...
-            break;
+        }
+        else if (const auto* keyPressed = event->getIf<sf::Event::KeyPressed>())
+        {
+            if (keyPressed->scancode == sf::Keyboard::Scancode::Escape)
+                window.close();
+        }
     }
+
+    // Remainder of main loop
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:46]:
```suggestion
    // Rest of the main loop
```

File: pages/tutorials/3.0/window/events.md [hunk 9]
```diff
@@ -7,259 +7,329 @@
...
+{
+    if (keyPressed.scancode == sf::Keyboard::Scancode::Escape)
+        window.close();
+};
+
+while (window.isOpen())
+{
+    window.handleEvents(onClose, onKeyPressed);
+
+    // Remainder of main loop
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:82]:
```suggestion
    // Rest of the main loop
```

File: pages/tutorials/3.0/window/events.md [hunk 10]
```diff
@@ -7,259 +7,329 @@
...
+}
+```
 
 ## The Closed event
 
-The `sf::Event::Closed` event is triggered when the user wants to close the window, through any of the possible methods the window manager provides ("close" button, keyboard shortcut, etc.). This event only represents a close request, the window is not yet closed when the event is received.
+The `sf::Event::Closed` event is triggered when the user wants to close the window through any of the possible methods the window manager provides ("close" button, keyboard shortcut, etc.).
+This event only represents a close request.
+The window is not yet closed when the event is received.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:90]: I feel like those two sentences only make sense when they're fully connected.

```suggestion
This event only represents a close request, meaning the window is not yet closed when the event is received.
```

File: pages/tutorials/3.0/window/events.md [hunk 11]
```diff
@@ -7,259 +7,329 @@
...
 
-Typical code will just call `window.close()` in reaction to this event, to actually close the window. However, you may also want to do something else first, like saving the current application state or asking the user what to do. If you don't do anything, the window remains open.
+Typically, one will just call `window.close()` in reaction to this event to actually close the window.
+However, you may also want to do something else first like saving the current application state or asking the user what to do.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:93]: I trust your comma skills, but to me there's a clear pause here... 😅 

```suggestion
However, you may also want to do something else first, like saving the current application state or asking the user what to do.
```

File: pages/tutorials/3.0/window/events.md [hunk 12]
```diff
@@ -7,259 +7,329 @@
...
+If you don't do anything, the window remains open.
 
-There's no member associated with this event in the [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") union.
+There's no data associated with this event.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:96]:
```suggestion
There's no event data associated with this event.
```

File: pages/tutorials/3.0/window/events.md [hunk 13]
```diff
@@ -7,259 +7,329 @@
...
 ```
 
 ## The Resized event
 
 The `sf::Event::Resized` event is triggered when the window is resized, either through user action or programmatically by calling `window.setSize`.
-
 You can use this event to adjust the rendering settings: the viewport if you use OpenGL directly, or the current view if you use sfml-graphics.
 
-The member associated with this event is `event.size`, it contains the new size of the window.
+There's no data associated with this event.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:108]: Yes there is 😄 

```suggestion
The member associated with this event is `event.size`, it contains the new size of the window.
```

File: pages/tutorials/3.0/window/events.md [hunk 14]
```diff
@@ -7,259 +7,329 @@
...
-
-The `sf::Event::LostFocus` and `sf::Event::GainedFocus` events are triggered when the window loses/gains focus, which happens when the user switches the currently active window. When the window is out of focus, it doesn't receive keyboard events.
+## The FocusLost and FocusGained events
 
+The `sf::Event::FocusLost` and `sf::Event::FocusGained` events are triggered when the window loses/gains focus, which happens when the user switches the currently active window.
+When the window is out of focus, it doesn't receive keyboard events.
 This event can be used e.g. if you want to pause your game when the window is inactive.
 
-There's no member associated with these events in the [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") union.
+There's no data associated with this event.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:124]:
```suggestion
There's no event data associated with this event.
```

File: pages/tutorials/3.0/window/events.md [hunk 15]
```diff
@@ -7,259 +7,329 @@
...
-The member associated with this event is `event.size`, it contains the new size of the window.
+There's no data associated with this event.
 
 ```cpp
-if (event.type == sf::Event::Resized)
+if (event->is<sf::Event::Resized>())
 {
     std::cout << "new width: " << event.size.width << std::endl;
     std::cout << "new height: " << event.size.height << std::endl;
 }
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:115]: Needs to be changed

File: pages/tutorials/3.0/window/events.md [hunk 16]
```diff
@@ -7,259 +7,329 @@
...
+The `sf::Event::TextEntered` event is triggered when a character is typed.
+This must not be confused with the `KeyPressed` event.
+`TextEntered` interprets the user input and produces the appropriate printable character.
+For example, pressing '^' then 'e' on a French keyboard will produce two `KeyPressed` events, but a single `TextEntered` event containing the 'ê' character.
+It works with all the input methods provided by the operating system, even the most specific or complex ones.
 
 This event is typically used to catch user input in a text field.
 
-The member associated with this event is `event.text`, it contains the Unicode value of the entered character. You can either put it directly in a [`sf::String`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1String.php "sf::String documentation"), or cast it to a `char` after making sure that it is in the ASCII range (0 - 127).
+You can either put this data directly in a [`sf::String`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1String.php "sf::String documentation"), or cast it to a `char` after making sure that it is in the ASCII range (0 - 127).
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:144]: We can't just drop the most important information, i.e. how to access the data and that it's an Unicode value.

```suggestion
The member associated with this event is `textEntered.text`, it contains the Unicode value of the entered character.
You can either put this data directly in a [`sf::String`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1String.php "sf::String documentation"), or cast it to a `char` after making sure that it is in the ASCII range (0 - 127).
```

File: pages/tutorials/3.0/window/events.md [hunk 17]
```diff
@@ -7,259 +7,329 @@
...
 }
 ```
 
-Note that, since they are part of the Unicode standard, some non-printable characters such as _backspace_ are generated by this event. In most cases you'll need to filter them out.
+Note that since they are part of the Unicode standard, some non-printable characters such as _backspace_ are generated by this event.
+In most cases you'll need to filter them out.
 
-Many programmers use the `KeyPressed` event to get user input, and start to implement crazy algorithms that try to interpret all the possible key combinations to produce correct characters. Don't do that!
+Many programmers use the `KeyPressed` event to get user input and start to implement crazy algorithms that try to interpret all the possible key combinations to produce correct characters.
+Don't do that!
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:158]:
```suggestion
!!! warning

    Many programmers use the `KeyPressed` event to get user input and start to implement crazy algorithms that try to interpret all the possible key combinations to produce correct characters.
    Don't do that!
```

File: pages/tutorials/3.0/window/events.md [hunk 18]
```diff
@@ -7,259 +7,329 @@
...
 
 ## The KeyPressed and KeyReleased events
 
 The `sf::Event::KeyPressed` and `sf::Event::KeyReleased` events are triggered when a keyboard key is pressed/released.
 
-If a key is held, multiple `KeyPressed` events will be generated, at the default operating system delay (ie. the same delay that applies when you hold a letter in a text editor). To disable repeated `KeyPressed` events, you can call `window.setKeyRepeatEnabled(false)`. On the flip side, it is obvious that `KeyReleased` events can never be repeated.
+If a key is held, multiple `KeyPressed` events will be generated, at the default operating system delay (ie. the same delay that applies when you hold a letter in a text editor).
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:164]:
```suggestion
If a key is held, multiple `KeyPressed` events will be generated, at the default operating system delay (i.e. the same delay that applies when you hold a letter in a text editor).
```

File: pages/tutorials/3.0/window/events.md [hunk 19]
```diff
@@ -7,259 +7,329 @@
...
 }
 ```
 
-Note that some keys have a special meaning for the operating system, and will lead to unexpected behavior. An example is the F10 key on Windows, which "steals" the focus, or the F12 key which starts the debugger when using Visual Studio.
-
-## The MouseWheelMoved event
-
-The `sf::Event::MouseWheelMoved` event is **deprecated** since SFML 2.3, use the MouseWheelScrolled event instead.
+Note that some keys have a special meaning for the operating system, and will lead to unexpected behavior.
+An example is the F10 key on Windows, which "steals" the focus, or the F12 key which starts the debugger when using Visual Studio.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:206]:
```suggestion
!!! note

    Note that some keys have a special meaning for the operating system, and will lead to unexpected behavior.
    An example is the F10 key on Windows, which "steals" the focus, or the F12 key which starts the debugger when using Visual Studio.
```

File: pages/tutorials/3.0/window/events.md [hunk 20]
```diff
@@ -7,259 +7,329 @@
...
+        std::cout << "mouse x: " << mouseButtonPressed->position.x << std::endl;
+        std::cout << "mouse y: " << mouseButtonPressed->position.y << std::endl;
     }
 }
 ```
 
 ## The MouseMoved event
 
 The `sf::Event::MouseMoved` event is triggered when the mouse moves within the window.
+This event is triggered even if the window isn't focused.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:251]: I think it's okay to repeat it:

```suggestion
The `sf::Event::MouseMoved` is triggered even if the window isn't focused.
```

File: pages/tutorials/3.0/window/events.md [hunk 21]
```diff
@@ -7,259 +7,329 @@
...
+However, it is triggered only when the mouse moves within the inner area of the window, not when it moves over the title bar or borders.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:252]:
```suggestion

!!! note

    This event is triggered even if the window isn't focused, but only when the mouse moves within the inner area of the window, not when it moves over the title bar or window borders.
```

File: pages/tutorials/3.0/window/events.md [hunk 22]
```diff
@@ -7,259 +7,329 @@
...
+    std::cout << "new mouse y: " << mouseMoved->position.y << std::endl;
+}
+```
+
+## The MouseMovedRaw event
+
+The `sf::Event::MouseMovedRaw` event is triggered when the mouse moved within the window even if the mouse is moved a distance too small to perceive.
 
-This event is triggered even if the window isn't focused. However, it is triggered only when the mouse moves within the inner area of the window, not when it moves over the title bar or borders.
+While the MouseMoved position value is dependent on the screen resolution, raw data is not.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:268]:
```suggestion
While the `sf::Event::MouseMoved` position value is dependent on the screen resolution, the raw data of this event is not.
```

File: pages/tutorials/3.0/window/events.md [hunk 23]
```diff
@@ -7,259 +7,329 @@
...
+If the physical mouse is moved too little to cause the screen cursor to move at least a single pixel, no MouseMoved event will be generated.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:269]: I think this needs to be moved up to `MouseMoved`

```suggestion
If the physical mouse isn't moved enough to cause the cursor on the screen to move at least a single pixel, no `MouseMoved` event will be generated.
```

File: pages/tutorials/3.0/window/events.md [hunk 24]
```diff
@@ -7,259 +7,329 @@
...
+if (const auto* mouseMovedRaw = event->getIf<sf::Event::MouseMovedRaw>())
 {
-    std::cout << "new mouse x: " << event.mouseMove.x << std::endl;
-    std::cout << "new mouse y: " << event.mouseMove.y << std::endl;
+    std::cout << "new mouse x: " << mouseMoved->delta.x << std::endl;
+    std::cout << "new mouse y: " << mouseMoved->delta.y << std::endl;
 }
 ```
 
+Currently, raw mouse input events will only be generated on Windows and Linux.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:282]: I guess this could either be a `!!! warning` or `!!! note`. Maybe we need to come up with a clearer definition when to use what.

File: pages/tutorials/3.0/window/events.md (hunk 22 above)
Comment (eXpl0it3r) [reply pages/tutorials/3.0/window/events.md:268]: I feel like this description isn't enough. We now just know what it is not, but we still don't know, what the data *is*.

File: pages/tutorials/3.0/window/events.md [hunk 25]
```diff
@@ -7,259 +7,329 @@
...
 
-if (event.type == sf::Event::MouseLeft)
+if (event->is<sf::Event::MouseLeft>())
     std::cout << "the mouse cursor has left the window" << std::endl;
 ```
 
 ## The JoystickButtonPressed and JoystickButtonReleased events
 
 The `sf::Event::JoystickButtonPressed` and `sf::Event::JoystickButtonReleased` events are triggered when a joystick button is pressed/released.
-
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:231]: I think this makes sense as a new paragraph.

File: pages/tutorials/3.0/window/events.md [hunk 26]
```diff
@@ -7,259 +7,329 @@
...
 SFML supports up to 8 joysticks and 32 buttons.
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:300]: I think this makes sense as a new paragraph.

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Naturally this got a few more comments.

General things we'll likely have to clarify on:
- When to use which [Admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/)
- How we differentiate between calling static functions and referencing member functions (e.g. `sf::JoyStick::setJoystickThreshold` vs `sf::WindowBase::pollEvent`)

And I believe we should look out for using "simple English" for most things. Most users are likely not native speakers.

File: pages/tutorials/3.0/window/events.md (hunk 26 above)
Comment (ChrisThrasher) [reply pages/tutorials/3.0/window/events.md:300]: It felt a bit awkward to have 3 sentences in a row all on their own paragraphs. Felt like we were adding whitespace for no good reason. I'll restore that newline though.

File: pages/tutorials/3.0/window/events.md (hunk 14 above)
Comment (ChrisThrasher) [reply pages/tutorials/3.0/window/events.md:124]: Are you sure? It feels awkward to repeat the word `event` twice.

File: pages/tutorials/3.0/window/events.md (hunk 11 above)
Comment (ChrisThrasher) [reply pages/tutorials/3.0/window/events.md:93]: Commas tend to represent a pause in one's speech but they're not defined as such. In this case I don't think the comma is correct. It may exist in a grey area where it's not necessarily correct nor incorrect to use it. English is a bit of a fuzzy language.

Comment (eXpl0it3r) [reply pages/tutorials/3.0/window/events.md:97]: Grammarly has a special rule for "such as", which one could use in place of "like".

image

Given that this a nonrestrictive clause, it would require a comma.

> English is a bit of a fuzzy language.

From a German perspective, that's a bit of an understatement. 😄  
The German language has a central and detailed grammatical definition of basically everything. Where English has countless different definitions.

File: pages/tutorials/3.0/window/events.md (hunk 14 above)
Comment (eXpl0it3r) [reply pages/tutorials/3.0/window/events.md:128]: Data is such a generic word, that I struggle with it, when it's not more explicitly defined.
It could be some random data attached to the event, or it could be data specifically related to the given event.
That's why I feel "event data" makes it clearer what kind of data we're talking about here.

File: pages/tutorials/3.0/window/events.md [hunk 27]
```diff
@@ -7,190 +7,252 @@
...
+The data associated with this event contains the number of ticks the wheel has moved, what the orientation of the wheel is and the current position of the mouse cursor.
 
 ```cpp
-if (event.type == sf::Event::MouseWheelScrolled)
+if (const auto* mouseWheelScrolled = event->getIf<sf::Event::MouseWheelScrolled>())
 {
-    if (event.mouseWheelScroll.wheel == sf::Mouse::VerticalWheel)
+    switch (mouseWheelScrolled->wheel)
+    {
+    }
```
Comment (eXpl0it3r) [inline pages/tutorials/3.0/window/events.md:228]: Looks like some coding artifact sneaked in here

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Just some tiny changes, looking pretty good otherwise!

*Look at me using a single review batch* 😉

**Review by eXpl0it3r — APPROVED** [review]
